### PR TITLE
feat(models,training)!: universal loss scalers for spectral losses

### PR DESCRIPTION
**Pull request by sahahner**
This pull request introduces several important updates to the spectral loss and scaling framework. The main changes are the addition of a new spectral dimension scaler, the depreciation of the `SpectralL2Loss`, introduction of a new `PowerSpectrumLoss`, and enhancements to ensure compatibility of scalers with spectral losses. Associated tests and imports are updated accordingly.

**Spectral Losses and Scalers:**
- Added new spectral dimension scaler: `SpectralDimensionScaler`, providing uniform weighted scaling over spectral dimensions. This means we do not depend on the workaround anymore to exclude the node scalers for the spectral losses.
- Introduced tests for the new spectral dimension scaler to verify correct behavior and formulas.

**Loss Function Updates:**
- Depercate `SpectralL2Loss`, as it is an expensive version of the MSE
- Introduce the `PowerSpectrumLoss`, which computes L2 loss on the power spectrum (energy per total wavenumber). All relevant imports, test cases, and references have been updated to use `PowerSpectrumLoss`. 

**Compatibility and Validation:**
- Added `_assert_spectral_scalers_compatible` to ensure that scalers applied on the grid dimension in spectral losses are correctly sized for the spectral domain, preventing mismatches between scaler and data shapes. 

## Description
Issue: https://github.com/ecmwf/anemoi-core/issues/967

## What problem does this change solve?
Introduce scalers for Spectral losses.

## What issue or task does this change relate to?
[](https://github.com/ecmwf/anemoi-core/issues/967)

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/Contributor-License-Agreement.md)

----
📚 Documentation preview 📚: https://anemoi-training--1253.org.readthedocs.build/en/1253/

----
📚 Documentation preview 📚: https://anemoi-graphs--1253.org.readthedocs.build/en/1253/

----
📚 Documentation preview 📚: https://anemoi-models--1253.org.readthedocs.build/en/1253/

### CODE REVIEW

File: training/docs/modules/losses.rst [hunk 1]
```diff
@@ -365,7 +366,7 @@ Gaussian-weighted nearest-neighbour weights.
            row_normalize: false
 
 Spectral kernel CRPS
---------------------
+====================
```
[the remark this answers is not in the document]
Comment (sahahner) [reply training/docs/modules/losses.rst:370]: adressed

File: training/src/anemoi/training/losses/scalers/spectral.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        self.spectral_dims = spectral_dims if spectral_dims is not None else self.n_spectral_modes
+
+    def get_scaling_values(self, **_kwargs) -> torch.Tensor:
+        """Return uniform scaling values (i.e. ones).
+
+        Returns
+        -------
+        torch.Tensor
+            Scaling values as a torch tensor.
+        """
```
[the remark this answers is not in the document]
Comment (sahahner) [reply training/src/anemoi/training/losses/scalers/spectral.py:72]: done

File: training/src/anemoi/training/losses/spectral.py [hunk 3]
```diff
@@ -415,16 +457,18 @@ def forward(
         is_sharded = channel_shard_sizes is not None
         group = group if is_sharded else None
 
-        pred_spectral = self._to_spectral_flat(pred)
-        target_spectral = self._to_spectral_flat(target)
+        sc_pred = self.transform.forward(pred)
+        sc_target = self.transform.forward(target)
+        pred_amp = self.transform.power_spectral_density(sc_pred)
+        target_amp = self.transform.power_spectral_density(sc_target)
+        diff = (pred_amp - target_amp) ** 2
```
[the remark this answers is not in the document]
Comment (sahahner) [reply training/src/anemoi/training/losses/spectral.py]: done

File: training/tests/unit/losses/test_loss_scaling.py [hunk 4]
```diff
@@ -557,3 +558,22 @@ def test_lead_time_decay_loss_scaling(
...
+    [(4, 4), (16, 16), (64, 64 * 64), (193, 193)],
+)
+def test_uniform_spectral_scaler(n_spectral_modes: int, spectral_dims: int) -> None:
+    """SpectralDimensionScaler: scaler values should all be identical (uniform)."""
+    scaler = SpectralDimensionScaler(n_spectral_modes=n_spectral_modes, spectral_dims=spectral_dims)
+    values = scaler.get_scaling_values()
+
+    assert values.shape == (spectral_dims,), f"Expected scaler to have shape ({spectral_dims},), got {values.shape}"
+    assert torch.allclose(values, values[0].expand_as(values)), f"Values are not uniform: {values}"
+    assert values[0] == 1.0 / n_spectral_modes, f"Expected values to be {1.0/n_spectral_modes}, got {values[0]}"
```
[the remark this answers is not in the document]
Comment (sahahner) [reply training/tests/unit/losses/test_loss_scaling.py]: done

File: training/src/anemoi/training/losses/scalers/spectral.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        LOGGER.info(
+            "Spectral Scaling: Applying %s with n_spectral_modes=%d.",
+            self.__class__.__name__,
+            self.n_spectral_modes,
+        )
+
+        return torch.ones(self.spectral_dims, dtype=torch.float32) / self.n_spectral_modes
```
Comment (mc4117) [inline training/src/anemoi/training/losses/scalers/spectral.py:79]: so the idea is to always have uniform scalers?

File: training/src/anemoi/training/losses/spectral.py [hunk 6]
```diff
@@ -511,22 +582,28 @@ def forward(
...
         # keeps the ratio bounded by Cauchy-Schwarz (up to numerical error)
         cross = torch.real(pred_spectral * torch.conj(target_spectral))
         denom = torch.sqrt(torch.abs(pred_spectral) ** 2 * torch.abs(target_spectral) ** 2 + eps)
         correlation = torch.clamp(cross / denom, min=-1.0, max=1.0)
 
-        # apply weighting/scaling after correlation is computed
-        result = (1 - correlation) / n_modes
-        result = self._reshard_spectral_loss(result, group, channel_shard_sizes, grid_dim)
+        # compute correlation
+        result = 1 - correlation
```
Comment (mc4117) [inline training/src/anemoi/training/losses/spectral.py:593]: so the /n_modes is no longer here because it's now in the scalers, right?

File: training/src/anemoi/training/losses/scalers/spectral.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+import torch
+
+from anemoi.training.losses.scalers.base_scaler import BaseScaler
+from anemoi.training.utils.enums import TensorDim
+
+LOGGER = logging.getLogger(__name__)
+
+
+class SpectralDimensionScaler(BaseScaler):
```
Comment (mc4117) [inline training/src/anemoi/training/losses/scalers/spectral.py:21]: Do you need to add SpectralDimensionScaler to the general scalers configs?

File: training/tests/unit/losses/test_loss_function.py [hunk 8]
```diff
@@ -614,39 +659,45 @@ def _octahedral_expected_points(nlat: int) -> int:
...
-    pred = torch.zeros((2, 1, 1, expected_points, nvars))
-    target = torch.zeros_like(pred)
-    _assert_variable_and_scalar_shapes(loss, pred, target, nvars=nvars)
+    # Loss should fail with wrong grid size
     pred_wrong = torch.zeros((2, 1, 1, expected_points + 1, nvars))
     target_wrong = torch.zeros_like(pred_wrong)
     with pytest.raises(AssertionError):
         _ = loss(pred_wrong, target_wrong, squash=True)
 
+    # fail for transform without PSD method (e.g. FFT2D)
```
Comment (mc4117) [inline training/tests/unit/losses/test_loss_function.py]: Small point but maybe put power spectral density method here?

File: training/src/anemoi/training/losses/scalers/spectral.py (hunk 5 above)
Comment (sahahner) [reply training/src/anemoi/training/losses/scalers/spectral.py:79]: That is the base scaler we have introduced in this PR. I have played around with scale-dependent scalers for the SHT-based losses in the previous PR, but that was not ready for main yet, so for now we are going for the uniform scalers only.

File: training/src/anemoi/training/losses/spectral.py (hunk 6 above)
Comment (sahahner) [reply training/src/anemoi/training/losses/spectral.py:593]: Yes, exactly. Normalisation over the number of modes should be done with the scalers from now on :)

File: training/src/anemoi/training/losses/scalers/spectral.py (hunk 7 above)
Comment (sahahner) [reply training/src/anemoi/training/losses/scalers/spectral.py:21]: Yes, we can add an example there. The spectral scaler must be adapted to the shape of the spectral projection, though, so I would leave it as a commented-out example.

**Review by mc4117 — APPROVED** [review]
LGTM!
